ls1intum/Artemis
# `Programming exercises`: Provide an overview to editors to generate code hints automatically

### Checklist
#### General
#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context
These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java.
This PR implements an overview of the whole code hint generation process including all required steps. In addition, the following additional functionality has been added:
- Edit the content of a solution entry
- Create new solution entries manually
- Option to only update the existing code hints (instead of recreating the hints including the deletion of old hints)

### Description
For the new overview page, I added a route using the current programming exercise from the route.

To save a (new) solution entry for an existing code hint within the process of saving the hint (and not the entry itself), we need to allow to set the solution entries (including test case and task) for an individual solution entry.


### Steps for Testing
Prerequisites:
- 1 Editor on Atlassian Stack (ts0, ts1, ts3, ts5)

1. Create a new programming exercise with the following configuration:
- Language: Java
- Project Type: Gradle
- Record testwise coverage: true
2. On the exercise details page, go to "Manage hints".
3. Click the button on the top right "Generate code hints".
4. Click through the different steps and make sure that everything works as expected. A later step should only be accessible if the previous step has been performed successfully.
5. Go to the third step and generate solution entries. Also create a manual entry.
6. Edit and delete some entries. Basically, test yourself through every button.
7. Go to the next step and press "Generate code hints". The hints should be displayed including the different solution entries.
8. Click "Re-create code hints" and verify that all existing hints have been deleted and only the "new" hints exist.
9. Click "Update code hints" and verify that all existing hints still exist with title, description, task, and content. But check that the existing hints have no associated solution entries.
10. Create a programming exercise as in step 1, but without the option to record the testwise coverage.
11. Verify that the whole process still works as expected, but without the first two steps (git-diff and coverage report).
12. Go again to "Manage hints". Then view and edit a hint. Make sure that no obvious errors occur.
13. Select an individual hint and click the button "Edit". Edit the content of some solution fragments, add new fragments, and remove existing fragments. Then save. Reopen the hint page and make sure that every fragment is as expected.

### Review Progress
#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| CodeHintService.java | 80% | 97% |
| ProgrammingExerciseResource.java | 81% | 99% |
| CodeHintResource.java | 50% | 85% |
| code-hint-generation-overview.component.ts | 76.2% | 97.3% |
| code-hint-generation-status.component.ts | 100% | 100% |
| manual-solution-entry-creation-modal.component.ts | 75% | 90.6% |
| solution-entry-details-modal.component.ts | 75% | 92.85% |
| code-hint-generation-step.component.ts | 58.3% | 82.14% |
| coverage-generation-step.component.ts | 75% | 88.89% |
| diff-generation-step.component.ts | 75% | 86.7% |
| solution-entry-generation-step.component.ts | 76.5% | 87.6% |
| exercise-hint-update.component.ts | 84.2% | 100% |
| exercise-hint-detail.component.ts | 80% | 100% |
| exercise-hint.component.ts | 68.2% | 91.3% |

### Screenshots
**Programming exercise details page:**
<img width="922" alt="Programming Exercise Details Page" src="https://user-images.githubusercontent.com/73833780/177231570-b3bb2729-21e3-4618-83a6-f96539f28702.png">

**Generate code hints button on "Manage hints" page:**
<img width="418" alt="Generate code hints button" src="https://user-images.githubusercontent.com/73833780/177231562-77e29f58-fa5c-43a4-9a77-d7190fd04797.png">

**Code Hint Management Step 1 (Git-Diff):**
<img width="1393" alt="CHM 1" src="https://user-images.githubusercontent.com/73833780/177231602-e5aef11e-9176-417c-acd6-79a726e6315b.png">

**Code Hint Management Step 2 (Coverage Report):**
<img width="1400" alt="CHM 2" src="https://user-images.githubusercontent.com/73833780/177231624-c70413b0-5f60-4fba-a50d-bc56af211c12.png">

**Code Hint Management Step 3 (Solution fragments/entries):**
<img width="1396" alt="CHM 3" src="https://user-images.githubusercontent.com/73833780/177366505-79ce1d9c-c9ed-4abf-93a5-8d938a9b7142.png">


**Code Hint Management Step 4 (Code Hints):**
<img width="1391" alt="CHM 4" src="https://user-images.githubusercontent.com/73833780/177231757-8d3c83e7-30d2-47b6-abad-dbb2b0c08b70.png">

**Edit a hint individually** (reached from "Manage Hints"-Page):
<img width="925" alt="Edit hint individually" src="https://user-images.githubusercontent.com/73833780/177338061-06a0a227-8a77-4fab-888c-15c768d953c3.png">
